### DISCUSSION TITLE
Why is it required to provide a s3, azure blob or github blob storage for migrations? (bbs2gh)

### DISCUSSION THREAD

**Discussion by ryaugusta** [github/gh-gei#1303] [Q&A] [answered]
I'm working on a migration where we are already exporting the .tar file from the BitBucket repo via the API. I want to use this repo tar file using `gh bbs2gh migrate-repo --archive-path ...`. I'm providing the tar file locally so that it can be uploaded during the migration step of the GEI. When I give the --archive-path I still run into an error saying I must specify external storage to continue.

Why do I need to provide a storage option? Can the GEI just use the local tar from the machine where the GEI is running without needing to upload to storage?

**Comment by dylan-smith** [comment, marked as answer]
The migration happens on the GitHub backend, this CLI is just a wrapper for various API calls. When the backend migration logic runs it needs access to the migration archive, and can’t reach out to your local machine to get it.

**Reply by ryaugusta** [reply]
Thanks @dylan-smith ; I assume then that the reason we need to provide `BBS_USERNAME`, `BBS_PASSWORD` and other bitbucket parameters after we already extract the tar is that the GEI will reach into bitbucket and extract the repository metadata info like PRs, Issues, Settings etc?

**Reply by dylan-smith** [reply]
No, you shouldn’t need to provide username and password if you provide the archive directly. In fact the CLI will error if you attempt to do so via args (I think).

**Reply by ryaugusta** [reply]
Hi @dylan-smith - I actually am not experiencing that..

I'm giving the archive directly; the GEI is still asking for the Username/Password for BBS..

```bash
gh bbs2gh migrate-repo --archive-path=~/Downloads/<repository>.tar --github-org=<org> --github-repo=<repository> --bbs-server-url=<bbs-server-url> --bbs-project=project --bbs-repo=sample-migration-repo --use-github-storage
```

Reponse:
`[2024-12-03 14:48:28] [ERROR] BBS_USERNAME environment variable is not set`

If this is not expected behavior as you're implying, I can open an issue for this..

**Reply by ryaugusta** [reply]
in this case, it's asking for me to set the `BBS_USERNAME` ..not unset it.

**Reply by dylan-smith** [reply]
thanks for the info, I can repro this and it looks like a bug to me. What I think happened is bbs-server-url used to be optional, and we had the logic that if you provided a server-url you must also provide username/password, but if you didn't provide a server-url then username/password were not required.

However, sometime back we realized that in order to process the BBS archive we need the server url (even if you are providing the archive directly via archive-path). However, the arg validation logic wasn't properly updated. You should be able to workaround it by providing any dummy values to get past the validation. And I'll pass this on to the engineering team to file a bug.

**Reply by ryaugusta** [reply]
Thanks @dylan-smith, passing in dummy creds starts the migration. If I could bug you about one more thing.

I'm seeing an error I haven't seen before with the gei, it gets to the `Importing Archive` step and is in progress for a while, which is expected, but then fails after some time with:

```bash
...
..."state":"FAILED","warningsCount":0,"failureReason":"Git source migration failed. Error message: A critical error occurred in the clone step of the git migration.","repositoryName":"ci-image-java-develop"}}}
...
[ERROR] OctoshiftCLI.OctoshiftCliException: Git source migration failed. Error message: A critical error occurred in the clone step of the git migration.
```
Any idea what would cause that clone to fail? I dug into the logs and there's no other message along with that. If this step is trying to clone the repository from GitHub which is being migrated it would make sense that it's failing because the migration isn't complete yet so the repo is in a "Preparing your repository" (or similar) state.

This morning I got a little more info on the error saying:

`[2024-12-04 08:34:39] [ERROR] Git source migration failed. Error message: A critical error occurred in the Clone Source step of the git migration.`

"Clone source" sounds like trying to clone the BBS repository.

**Reply by ryaugusta** [reply]
@dylan-smith  -- ignore the above, this is actually expected behavior on my clients security side.

One thing though, why can't we use a bearer authentication type with BBS.. reading through it seems like we can only use username/password. We have a restriction that we need to use Bearer so similarly to `GH_PAT` can we use `BBS_PAT`?

**Reply by dylan-smith** [reply]
The BBS API to generate a migration archive only supports username/password authentication. Super weird and annoying, but it's something you'd need to take up with Atlassian.

**Reply by ryaugusta** [reply]
Agh, more hurdles. Thanks again for all the information!
